x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/jwt/JwtRealmSettings.java

```diff
@@ -185,16 +170,13 @@ private JwtRealmSettings() {}
         )
     );
 
-    public static final Setting.AffixSetting<String> JWKSET_PATH = RealmSettings.simpleString(
+    public static final Setting.AffixSetting<String> JWKSET_PKC_PATH = RealmSettings.simpleString(
         TYPE,
         "jwkset_path",
         Setting.Property.NodeScope
     );
 
-    public static final Setting.AffixSetting<SecureString> ISSUER_HMAC_SECRET_KEY = RealmSettings.secureString(
-        TYPE,
-        "issuer_hmac_secret_key"
-    );
+    public static final Setting.AffixSetting<SecureString> JWKSET_HMAC_CONTENTS = RealmSettings.secureString(TYPE, "issuer_hmac");
```

Comment:
HMAC design changed to JWKSet. I am wondering if both settings need a name change. Possible examples:
- **jwkset_hmac**; instead of **issuer_hmac_secret_key**?
- **jwkset_pkc_path**; instead of **jwkset_path**?

Reply:
Why HMAC JWKSet?
1. You mentioned different HMAC per OIDC RP; OIDC realm has only one RP, but JWT realms can have many RPs.
2. Same code works for any JWKSet, no need for different code paths. Examples:
- JwtRealmsTest.java line 368 -> one-liner inputs list of HMAC/RSA/EC sig algs, returns list of corresponding JWKs
- JwtRealmsTest.java line 387 -> one-liner encodes HMAC JWKSet, for loading into ES keystore
- JwtRealmsTest.java line 388 -> one-liner encodes PKC JWKSet, for hosting on local filesystem or remote HTTPS server

Reply:
> HMAC design changed to JWKSet.

So we are expecting that folks will create a JWKSet from their HMAC key(s) and then pass it as a string to the configuration in the elasticsearch keystore ? Elasticsearch Secure Settings can hold files too, and folks are usually treating JWKSets as files ( and we also do this for the public key based JWKSets) , so could if be worth adding this as a file in the keystore instead of a string ?

 
> You mentioned different HMAC per OIDC RP; OIDC realm has only one RP, but JWT realms can have many RPs.

Agreed, but as I also mentioned in that discussion, I think that the vast majority of use cases for this realm will only have a single RP and a single HMAC key and I think we are hugely complicating the configuration of this use case by mandating that the HMAC key can only be configured as a JWKSet

> Same code works for any JWKSet, no need for different code paths. Examples:

I don't think that `one-liner` vs an `if-else` or similar in our _test_ code, is a significant argument in favor of selecting an approach.

My personal preference is to have two settings: 

1. an hmac key as string in the elasticsearch keystore
2. an hmac jwkset as a file in the elasticsearch keystore. 

We can validate that only one of the two is set for a realm. The first can be used when there is only a single HMAC key ( majority of cases). The latter can be used when there are multiple RPs with multiple HMAC keys talking to elasticsearch ( or even when there is one but the administrator has a JWKSet or prefers it ) 

I also believe we still need a CLI utility of sorts (in a separate PR) to create a JWKSet from one/multiple HMAC keys as I don't think that many users will have a JWKSet for their HMAC keys or a way to generate these.

Reply:
For HMAC format, both options need to be loaded from a file into Elasticsearch keystore. I don't see a difference there.

JWKSet has additional benefits besides multiple keys. JWKs support attributes like "kid", "usage", etc. The JWK loop takes advantage of the kid. I assume JWSVerifier takes advantage of usages. At a minimum, the kid check speeds up validation when multiple RSA/EC keys are in the PKC JWKSet. If uses same format, it benefits from the same optimizations.

Reply:
This is my view of how customer can ramp on JWT. It involves a CLI. I mentioned this in our weekly team meeting today, so it sounds like we had the same idea. Here is my idea.

CLI 1 generates up to three JWKSet files. Command line input is a CSV list of signature algorithms, so very easy to use.
1. **Issuer JWKSet file** - RSA/EC key pairs and HMAC secret keys (single or separate files, single might be easier)
2. **Realm PKC JWKSet file** - RSA/EC public keys (hosted in local file or remote HTTPS URL)
3. **Realm HMAC JWKSet file** - HMAC secret keys (hosted in ES keystore)

CLI 2 generates a signed JWT. Command line input is the issuer JWKSet file from CLI 1, key/value claims for the JWT, and an optional kid for the JWT header if they want to use a specific RSA vs EC vs HMAC key from the JWKSet. The JWKSet is clear test JSON, so they can view it to see the kids we can generate in CLI 1.

Overall, the customer has a working JWT issuer, and keys to use in their JWT realm. They can use the test JWT in a curl command to validate that they configured the JWT realm correctly. That test is independent of their bespoke JWT issuer or external OIDC workflow, so I think it will prevent a lot of SDHs about JWT realm misconfigurations.

In summary, my view is JWKSets are just KeyStores/TrustStores. Our PKI CLIs uses PKCS12 encrypted format, and JWT realm can use JWKSet cleartext format.

Reply:
> For HMAC format, I think both options you mentioned need to be loaded from a file into Elasticsearch keystore. I don't see a difference if we load raw bytes from a file, or a cleartext JWKSet from a file.

My suggestion is not for our benefit, but for the benefit of our users. 

> JWKSet has additional benefits besides multiple keys. JWKs support attributes like "kid", "usage", etc. The JWK loop takes advantage of the kid. I assume JWSVerifier takes advantage of usages. At a minimum, the kid check speeds up validation when multiple RSA/EC keys are in the PKC JWKSet. If uses same format, it benefits from the same optimizations.

I get that it makes our lives easier. My argument is that it makes the life of the administrator setting up this realm for the average use case considerably harder and thus I'd want us to reconsider. 

> Here is my idea.

I think that your idea makes a number of assumptions:
a) Users don't have keys already. This might be the case for arbitrary JWT signing clients, but most probably not the case for OIDC OPs.
b) OP ( or JWT signing clients in general ) are happy/willing/allowed to use our CLI to generate keys for the JWT signing, 
c) OPs ( or JWT signing clients in general ) can be configured with JWKSets and their administrators are willing to do so , instead of using the raw keys files and have their software generate the JWKs on runtime, 

> CLI 2 generates a signed JWT.

I won't necessarily object this idea but it goes a bit away from the scope of this piece of work, and our functionality. We don't usually create specific assisting CLI Tools.

> In summary, my view is JWKSets are just KeyStores/TrustStores. Our PKI CLIs uses PKCS12 encrypted format, and JWT realm can use JWKSet cleartext format.

In this analogy though, we don't mandate that users only use PKCS12 containers for their key data. We support PEM files too and this is more or less what I'm asking we do here. 

> it really simplifies the customer workflow for setting up and verifying their JWT realm is configured OK.

I disagree with this. My view is that when a user has a secret key they need to use (which I believe is the most common use case we'll be seeing), the easiest way to set this up is to configure this value both in elasticsearch and the OP/JWT signer. It's just simpler than figuring out that there is a CLI tool you need to use and pass this secret key so that it will transform it to another format that you can then configure in elasticsearch and the OP/JWT signer. 

I'm not arguing against JWKSets in general, I get the benefits. The only thing I am asking for is that we allow an easy way for users to configure the simplest/easier/most common(granted that's anecdotal) use case, in addition to that.

Reply:
Code is refactored.